### PR TITLE
socket_zep: fix reboot

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The `_zep_params_setup()` function will modify the `argv` string passed to it.
This is a problem because that string is re-used on reboot.

The modified string is then later processed in `socket_zep_setup()`, so we have to keep the memory around.

~~The `strdup()` fulfills all this with the only downside that we now have a minor memory leak on every native reboot with ZEP.~~

~~IMHO as this is a pretty rare event and on `native` there is plenty of memory, a few bytes lost are not a big deal.~~

Reboot will call `execve` so all memory is freed.

### Testing procedure

On a `native` instance with `socket_zep`, type `reboot`.

#### master

```
reboot


		!! REBOOT !!

usage: /home/benpicco/dev/RIOT/examples/gnrc_border_router/bin/native/gnrc_border_router.elf <tap interface 1> [-i <id>] [-d] [-e|-E] [-o] [-c <tty>]
 -z [[<laddr>:<lport>,]<raddr>:<rport>]
 help: /home/benpicco/dev/RIOT/examples/gnrc_border_router/bin/native/gnrc_border_router.elf -h


Options:
    -h, --help
        print this help message
    -i <id>, --id=<id>
        specify instance id (set by config module)
    -s <seed>, --seed=<seed>
        specify srandom(3) seed (/dev/random is used instead of random(3) if
        the option is omitted)
    -d, --daemonize
        daemonize native instance
    -e, --stderr-pipe
        redirect stderr to file
    -E, --stderr-noredirect
        do not redirect stderr (i.e. leave sterr unchanged despite
        daemon/socket io)
    -o, --stdout-pipe
        redirect stdout to file (/tmp/riot.stdout.PID) when not attached
        to socket
    -c <tty>, --uart-tty=<tty>
        specify TTY device for UART. This argument can be used multiple
        times (up to UART_NUMOF)
    -z [<laddr>:<lport>,]<raddr>:<rport> --zep=[<laddr>:<lport>,]<raddr>:<rport>
        provide a ZEP interface with local address and port (<laddr>, <lport>)
        and remote address and port (default local: [::]:17754).
        Required to be provided SOCKET_ZEP_MAX times
make: *** [/home/benpicco/dev/RIOT/examples/gnrc_border_router/../../Makefile.include:728: term] Error 1
```

#### this PR

```
reboot


		!! REBOOT !!

RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

gnrc_uhcpc: Using 6 as border interface and 7 as wireless interface.
uhcp_client(): sending REQ...
got packet from fe80::2 port 12345
got packet from fe80::64e8:efff:fe02:ec3c port 48265
uhcp: push from fe80::64e8:efff:fe02:ec3c:48265 prefix=2001:db8::/64
gnrc_uhcpc: uhcp_handle_prefix(): add compression context 0 for prefix 2001:db8::25a:4550:a00:de49/64
gnrc_uhcpc: uhcp_handle_prefix(): configured new prefix 2001:db8::25a:4550:a00:de49/64
main(): This is RIOT! (Version: 2020.10-devel-715-g538fa-examples/gnrc_border_router-native)
RIOT border router example application
All up, running the shell now
```

### CODE REVIEW

**Comment by miri64** [comment]
> The `strdup()` fulfills all this with the only downside that we now have a minor memory leak on every native reboot with ZEP.

Erm.... do not like oO

**Review by miri64 — CHANGES_REQUESTED** [review]
`strdup()` could be moved to `_zep_params_setup()` and the result of it freed after that function is done. Also some explanatory comment, why `strdup()` is done would be nice :-).

**Comment by benpicco** [comment]
> `strdup()` could be moved to `_zep_params_setup()` and the result of it freed after that function is done

Unfortunately this doesn't work as `_zep_params_setup()` will just split the string, the actual parsing happens in `socket_zep_setup()`.

Now the proper fix would be to move the parsing to `_zep_params_setup` and not pass strings around.

**Comment by miri64** [comment]
Why wouldn't this work?

```diff
diff --git a/cpu/native/startup.c b/cpu/native/startup.c
index 4c29e619d9..7df852c6aa 100644
--- a/cpu/native/startup.c
+++ b/cpu/native/startup.c
@@ -371,6 +371,8 @@ static void _zep_params_setup(char *zep_str, int zep)
 {
     char *save_ptr, *first_ep, *second_ep;
 
+    /* don't destroy original string to re-use it at reboot */
+    zep_str = strdup(zep_str);
     if ((first_ep = strtok_r(zep_str, ",", &save_ptr)) == NULL) {
         usage_exit(EXIT_FAILURE);
     }
@@ -387,6 +389,7 @@ static void _zep_params_setup(char *zep_str, int zep)
         _parse_ep_str(second_ep, &socket_zep_params[zep].remote_addr,
                       &socket_zep_params[zep].remote_port);
     }
+    free(zep_str);
 }
 #endif
 
@@ -495,7 +498,7 @@ __attribute__((constructor)) static void startup(int argc, char **argv, char **e
 #endif
 #ifdef MODULE_SOCKET_ZEP
             case 'z':
-                _zep_params_setup(strdup(optarg), zeps++);
+                _zep_params_setup(optarg, zeps++);
                 break;
 #endif
 #ifdef MODULE_PERIPH_SPIDEV_LINUX
```

This also would have the benefit, that the concerns are kept to the function that needs to duplicate the string, not the caller.

**Comment by benpicco** [comment]
Only if we also change `socket_zep_params_t` to not contain pointers to that string.

**Comment by miri64** [comment]
> Only if we also change `socket_zep_params_t` to not contain pointers to that string.

Ah... then why keep the pointer(s) globally and remove them once they _are_ used.

> Now the proper fix would be to move the parsing to `_zep_params_setup` and not pass strings around.
> But as I wasn't even sure if ZEP would help with my problem, I was a bit lazy and looked for a quick solution ;)

I would prefer to find a simpler solution, as it would bunch together separate concerns. Simplicity is also why I would not like to use `execve`.

**Comment by benpicco** [comment]
> Simplicity is also why I would not like to use `execve`.

But `native` *does* use `execve` to implement reboot. That's the reason we need to keep around the original `argv`.

I think moving the parsing to `_zep_params_setup()` and then passing around a `struct addrinfo *` instead of a bunch of strings will be the best solution here.

**Comment by miri64** [comment]
> But `native` _does_ use `execve` to implement reboot. That's the reason we need to keep around the original `argv`.

Then can't we free `zep_str` after `execve` was called?

> I think moving the parsing to `_zep_params_setup()` and then passing around a `struct addrinfo *` instead of a bunch of strings will be the best solution here.

But that would add the dependencies required for address parsing to `startup.c`.

**Comment by benpicco** [comment]
> Then can't we free `zep_str` after `execve` was called?

No because *„This causes the program that is currently being run by the calling process to be replaced with a new program“*.
The heap is automatically freed as a result of that.

**Comment by miri64** [comment]
Arghs, sorry. I wasn't reading correctly. I thought you wanted to do some black magic with `execve` and did not notice that you adapted OP.

File: cpu/native/startup.c [hunk 1]
```diff
@@ -495,7 +495,7 @@ __attribute__((constructor)) static void startup(int argc, char **argv, char **e
 #endif
 #ifdef MODULE_SOCKET_ZEP
             case 'z':
-                _zep_params_setup(optarg, zeps++);
+                _zep_params_setup(strdup(optarg), zeps++);
```
Comment (miri64) [inline cpu/native/startup.c:498]: Still would prefer if the `strdup` was moved into `_zep_params_setup()`.

Comment (miri64) [reply cpu/native/startup.c:498]: (with a comment)

Comment (benpicco) [reply cpu/native/startup.c]: Ok, moved it there and amended the commit message too.

**Review by miri64 — APPROVED** [review]
ACK. I built the `gnrc_border_router` example for `native`

```sh
BOARD=native make -C examples/gnrc_border_router/ -j
```

and ran it. I confirmed the behavior in both `master` and this PR. Using the `all-asan` target to also check for leaks, but was not able to find any. `all-valgrind`/`term-valgrind` did not work, as `term-valgrind` does not handle the `TERMFLAGS` properly :-/.

**Pull request merged**